### ISSUE TITLE
what's this io_tools?

### ISSUE THREAD

**Issue by lucasjinreal** [bebuch/tar#4]
what's this io_tools?

**Comment by bebuch** [comment]
https://github.com/bebuch/io_tools

My libs on Github are outdated, so be careful when using them! For the past few years, I have maintained my libraries on our company's internal server. However, tar has not received any further updates during this time, so this is still the current version.

Feel free to fork them if you want to develop them further.

**Comment by lucasjinreal** [comment]
@bebuch  thank u for your work! I am currently trying to using a very lightweighted parsing tar package lib for me to load tvm models.

However, I am really know a very tiny of any tar protocals or C++ deep, I make your code running but seems can not parse tar packages correctly. I wonder:

1) Do u know what's this error going on? https://github.com/bebuch/tar/issues/5
2) I uncommented out these check lines: 

```c++
//   if (magic != "ustar") {
//     throw std::runtime_error(
//         "Tar: loaded file without magic 'ustar', magic is: '" +
//         thor::io::mask_non_print(magic) + "'");
    
//   }

//   if (checksum != calc_checksum(buffer)) {
//     throw std::runtime_error("Tar: loaded file with wrong checksum");
//   }

  return std::make_tuple(std::move(filename),
                         static_cast<std::size_t>(std::stol(size, 0, 8)));
```
since I think it maybe these magic number changes for am using ubuntu 20.10 and it's decent. But I still got errors:

```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  stol

```
Any idea for thoese errors?

3) Do u know, or  can you kindly suggest me some of these can work on parsing tar packages now?

thanks very much

**Comment by bebuch** [comment]
If I remember it correctly, there exist different versions of TAR archives. My lib can only handle the simple standard version without any extensions. Take a look here:

https://en.wikipedia.org/wiki/Tar_(computing)#UStar_format

The ustar is part of the magic number (the first 8 bytes of the file), which is listed in the TAR info box at the beginning of the Wikipedia article. This box also contains the hint "Magic number: [...] ustar [...] absent in pre-POSIX versions". You probably try to read an TAR archive that is written by a very very old TAR library or a TAR library which is not POSIX conform.

The TAR files written with my lib can be opened by the lib and the very most compression archive applications. On the other hand, if you create an TAR archive by another program you want to read with my lib, you need to make sure it is the one POSIX TAR archive format, my library supports.

Hope this helps. ;-)

**Comment by lucasjinreal** [comment]
@bebuch  thank u so much.

But, I am simply create by `tar -czvf test.tar test/` on Ubuntu 20.10, is that different or not POSXI?

**Comment by bebuch** [comment]
tl;dr Try `tar --format=posix -cvf test.tar test/` or `tar --format=gnu -cvf test.tar test/`.

(Note you must not flag `z` in `czvf`, just  `cvf`.)

----

Small correction, the Magic Number of a TAR archiv is not in the first 8 bytes, but at position 257 (Hex line 0x0100, second character; 0x0101 == 257).

You can use a Hex-Editor to check your TAR archiv. (On KDE I use Okteta, with Gnome you can use GHex if you don't want to install all the QT and KDE libraries Okteta depends on.)

----

https://manpages.ubuntu.com/manpages/bionic/en/man1/tar.1.html

Take a look at the section "Archive format selection". I'm not sure which one is the format I support, probably POSIX or GNU. Ubuntu uses GNU tar, so GNU is the default format on Ubuntu.

----

In any case, your mistake is a completely different one. You create a TAR archive, but then make a GNU zip compressed file from it, which contains the TAR archive but isn't one. The correct file extension for this would be `.tar.gz` or short `.tgz`. The `z` flag in `czvf` stands for Gnu Zip compression. ;-)

----

Just another note: I found at least two small bugs while studying my code 😅 If I have time, I will fix this in the next few days.

**Comment by lucasjinreal** [comment]
@bebuch You are with so bright eyes, looked at your so old code and can found 2 bugs. thanks for your help! I am able to load tar file with another implementation.

But may I ask how to save one file of them to local disk in C++?
They reads out as char content.

I merged it into my tiny repo: https://github.com/jinfagang/thor/blob/master/include/tar.h
